docs/proposals/registry/devfile-packaging.md

```diff
@@ -0,0 +1,51 @@
+
+<img width="742" alt="Screen Shot 2020-10-22 at 1 54 00 PM" src="https://user-images.githubusercontent.com/6880023/97219676-b2637200-17a0-11eb-8465-1063aa048768.png">
+
+
+Each folder under devfiles corresponds to a stack, with its own devfile and associated artifacts. There may be an index.json present for devfile consumers
+
+From this point, the Github repository may act as self-hosting devfile registry (as in the case of [odo-devfiles/registry](https://github.com/odo-devfiles/registry), or it may be built up into an Apache server container (as in the case of v1 registries like [eclipse/che-devfile-registry](https://github.com/eclipse/che-devfile-registry)).
+
+## Are any changes needed to accommodate OCI?
+Storing the devfiles (and any associated artifacts) on a dedicated Github repository (the “devfile registry repository”), is still a good idea.
+
+   - Provides a clear, straightforward and non-cluttered place for stack owners to contribute devfiles to
+   - Separates the devfile stacks from the OCI registry code and logic
+
+But, with the move to OCI registries, we need a way to easily package the devfile stacks from GitHub and load them into the OCI registry
+
+   - Needs to retrieve the devfiles without an internet connection
+     - Means we can’t have the metadata container git clone the devfiles before pushing them up
+   - Needs to be easily built and distributable by registry admins.  
+     - Once we support “deploying your own registry”, with custom devfiles this will become especially important
+   - Shouldn’t require building and deploying custom components of the registry to distribute (i.e. it shouldn’t be baked into the registry or metadata container images)
+     - Currently, the metadata container has the devfiles built into it, meaning users need to build their own metadata container image to customize devfiles
+
+## Proposed Changes
+
+**To address the requirements listed above, rather than bundling the devfiles as part of registry's bootstrap container, build the “devfile registry repository” into its own docker image**
+
+<img width="636" alt="Screen Shot 2020-10-26 at 3 39 33 PM" src="https://user-images.githubusercontent.com/6880023/97220180-767cdc80-17a1-11eb-85fd-cf1a5a623aeb.png">
+
+
+This container image will contain just the devfiles (and their artifacts), as well as any other required files.
+
+When deploying the OCI registry, the image will be specified as an init container, with an entrypoint command that copies the devfiles (and artifacts) to the bootstrap container’s volume. Then, as-is today, when the bootstrap container starts up, it generates the index.json and pushes the devfiles up into the OCI registry using `oras`.
```

Comment:
I'll refer the actual build of the devfile registry as "registry build" from now on.  This proposal defers the registry build including the generation of the index.json to the time of the deployment phase of the registry instead of the during the build phase in the registry life cycle. The problem that I see with this approach is that it is too late in the development/deployment cycle to try to do the registry build during registry deployment. The registry build includes the creation of the index.json and should be part of the on boarding process of the registry.  As part of the registry build, we'll need do any the validation of the stack and some of those validation may even potentially running any of the stack validation test. This process may take long so we should do it as part of the registry build phase instead of doing it in the deployment phase. 

Taking a step back, I still do not see the benefit of the proposed devfile only package instead of the current bootstrap image approach. At the end of day, it is the registry admin that needs to deploy the registry. There is always a build step to build from the registry repo source to the metadata container or the proposed repo content only image.  As long as we provide enough support for registry admin to easily build that final image, e.g. we provide a parent image including the build tools for that knows how to build the index.json and do the validation, the Dockerfile file for the bootstrap image should should remain fairly simple. Using this approach, the benefits are:
1. we can push all the heavy lifting job on the bootstrap image build which is done as part of the build phase of the registry image to shorten the time for deployment. 
2. Any validation problem with the content of the stacks will be found out during the build phase instead of delaying to find out in the deployment phase. and thus, reducing the risk of deployment.

Reply:
> I'll refer the actual build of the devfile registry as "registry build" from now on. This proposal defers the registry build including the generation of the index.json to the time of the deployment phase of the registry instead of the during the build phase in the registry life cycle. The problem that I see with this approach is that it is too late in the development/deployment cycle to try to do the registry build during registry deployment. The registry build includes the creation of the index.json and should be part of the on boarding process of the registry. As part of the registry build, we'll need do any the validation of the stack and some of those validation may even potentially running any of the stack validation test. This process may take long so we should do it as part of the registry build phase instead of doing it in the deployment phase.

So, that's a good point. I think my design proposal was a little ambiguous in this regard. I wasn't planning on offloading the registry build to the deployment, but I didn't realize the index.json generation was part of the "registry build". We can certainly generate the index.json as part of the registry build (as well as doing any necessary validation / testing), and then include that index.json in the "devfile image". 

> Taking a step back, I still do not see the benefit of the proposed devfile only package instead of the current bootstrap image approach. At the end of day, it is the registry admin that needs to deploy the registry. There is always a build step to build from the registry repo source to the metadata container or the proposed repo content only image. As long as we provide enough support for registry admin to easily build that final image, e.g. we provide a parent image including the build tools for that knows how to build the index.json and do the validation, the Dockerfile file for the bootstrap image should should remain fairly simple. Using this approach, the benefits are:

So, I still prefer keeping the devfiles / stacks separate from the bootstrap image, as it means that registry admins don't have to maintain (and keep up-to-date) separate forks of the bootstrap image just to bundle different devfiles in a registry. Especially for compatibility with new versions of the registry operator, it means that admins don't have to rebuild / rebase their custom metadata image just to ensure it works with the new operator.

What if we adapt the current proposal so that it's clear the "registry build" does the index.json generation, and any validation / testing as you mentioned, but it also builds up this "devfile image" for consumption (with the generated index.json)? Then I think all 3 of your points should be addressed, and we're able to keep the devfiles separate from the bootstrap container.

Reply:
Discussed on today's meeting:
- Can't avoid the registry admin having to update their "devfile image" between registry updates anyways, in case things such as the index.json or devfile packaging format change between releases
- My approach would require a second persistent volume
- Elson's approach is more sound and not any less complicated for the registry admin.

Reply:
@elsony Updated the design doc based on today's meeting, please re-review when you get the chance, thanks!